dvc/repo/reproduce.py

```diff
@@ -95,16 +95,21 @@ def reproduce(
         )
 
     ret = []
+    checked_stages = set()
     for target in targets:
-        stages = _reproduce_stages(active_graph, target, **kwargs)
+        stages, these_checked_stages = _reproduce_stages(
```

Comment:
What if we simply pass all targets to `_reproduce_stages` and do the duplication something like
```
pipeline = stages                                                               
if not single_item:                                                             
    for stage in stages:                                                        
        if downstream:                                                                    
            nodes = nx.dfs_preorder_nodes(G.copy().reverse(copy=False), stage)  
        else:                                                                   
            nodes = nx.dfs_postorder_nodes(G, stage)                            
        new = [node for node in nodes if node not in pipeline]                  
        pipeline.extend(new)                                                    
```

Reply:
Yep, It can definitely be de-duped in `_reproduce_stages`, I'll change that. I think there's a bug in the code above: the repro `target` stages will always be executed first, before their dependencies. I.e. if I specify D in 

                               E
                              / \
                             D   F
                            / \   \
                           B   C   G
                            \ /
                             A

Then my execution order would be [D, A, B, C]

Reply:
By the way, is it a design decision to avoid type hints? I'd usually use them but I can't see them elsewhere in the codebase

Reply:
> By the way, is it a design decision to avoid type hints? I'd usually use them but I can't see them elsewhere in the codebase

Not at all, we've just dropped py2 support in Jan and simply didn't use type hints widely yet. We do have some in stage-related code, IIRC. Please feel free to use them :slightly_smiling_face:

Reply:
> Then my execution order would be [D, A, B, C]

Good catch! Sorry, that was my smooth brain moment :slightly_smiling_face: